pkg/cache/gvk_caches.go

```diff
@@ -0,0 +1,173 @@
+limitations under the License.
+*/
+
+package cache
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/rest"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+)
+
+type GVKCachesOptions struct {
+	ObjectCacheFuncs map[client.Object]NewCacheFunc
+	DefaultCacheFunc NewCacheFunc
+}
+
+// GVKCachesBuilder - Builder function to create a new composite cache which contains
+// one or more caches based on a mapping of GVKs. Each GVK can be configured to
+// use a separate cache.
+func GVKCachesBuilder(gvkCachesOpts GVKCachesOptions) NewCacheFunc {
+	return func(config *rest.Config, opts Options) (Cache, error) {
+		opts, err := defaultOpts(config, opts)
+		if err != nil {
+			return nil, err
+		}
+
+		if gvkCachesOpts.DefaultCacheFunc == nil {
+			gvkCachesOpts.DefaultCacheFunc = New
+		}
+		defaultCache, err := gvkCachesOpts.DefaultCacheFunc(config, opts)
+		if err != nil {
+			return nil, err
+		}
+
+		gvkToCache := map[schema.GroupVersionKind]Cache{}
+		for obj, newCacheFunc := range gvkCachesOpts.ObjectCacheFuncs {
+			gvk, err := apiutil.GVKForObject(obj, opts.Scheme)
+			if err != nil {
+				return nil, err
+			}
+			gvkToCache[gvk], err = newCacheFunc(config, opts)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		return &gvkCaches{
+			gvkToCache:   gvkToCache,
+			defaultCache: defaultCache,
+			scheme:       opts.Scheme,
+		}, nil
+	}
+}
+
+type gvkCaches struct {
+	gvkToCache   map[schema.GroupVersionKind]Cache
+	defaultCache Cache
+
+	scheme *runtime.Scheme
+}
+
+func (c *gvkCaches) Get(ctx context.Context, key client.ObjectKey, obj client.Object) error {
+	gvk, err := apiutil.GVKForObject(obj, c.scheme)
+	if err != nil {
+		return err
+	}
+
+	if cache, ok := c.gvkToCache[gvk]; ok {
+		return cache.Get(ctx, key, obj)
+	}
+
+	return c.defaultCache.Get(ctx, key, obj)
+}
+
+func (c *gvkCaches) List(ctx context.Context, list client.ObjectList, opts ...client.ListOption) error {
+	gvk, err := apiutil.GVKForObject(list, c.scheme)
+	if err != nil {
+		return err
+	}
+
+	if !strings.HasSuffix(gvk.Kind, "List") {
+		return fmt.Errorf("non-list type %T (kind %q) passed as output", list, gvk)
+	}
+	gvk.Kind = strings.TrimSuffix(gvk.Kind, "List")
+
+	if cache, ok := c.gvkToCache[gvk]; ok {
+		return cache.List(ctx, list, opts...)
+	}
+	return c.defaultCache.List(ctx, list, opts...)
+}
+
+func (c *gvkCaches) GetInformer(ctx context.Context, obj client.Object) (Informer, error) {
+	gvk, err := apiutil.GVKForObject(obj, c.scheme)
+	if err != nil {
+		return nil, err
+	}
+
+	if cache, ok := c.gvkToCache[gvk]; ok {
+		return cache.GetInformer(ctx, obj)
+	}
+
+	return c.defaultCache.GetInformer(ctx, obj)
+}
+
+func (c *gvkCaches) GetInformerForKind(ctx context.Context, gvk schema.GroupVersionKind) (Informer, error) {
+	if cache, ok := c.gvkToCache[gvk]; ok {
+		return cache.GetInformerForKind(ctx, gvk)
```

Comment:
what in general would be the use case of having caches based on gvk? In the sense, what would be the scenario where it is more optimized than having caches for each namespace?

Reply:
The primary use case I can think of is this:
1. Primary object - global cache
2. Secondary resources - global cache or possibly dynamic multinamespace cache, depending on expected number of namespaces that would contain primary objects and RBAC permissions.
3. Controller configuration objects (possibly configmaps, secrets, or CRD config objects) - controller namespace

This solves two problems:
1. Memory savings: 
    - if using the dynamic multinamespace cache, an operator would only cache secondary objects in the namespaces where a CR is or has been present since the controller started
    - if a controller expects some if its objects to exist _only_ in the namespace it is running, it can watch that GVK in its namespace only, again saving cache space by not caching unrelated objects of the same GVK in other namespaces.
    - if a controller can trigger a reconciliation for its primary object based solely on the metadata of another object, it can setup a metadata-only cache, which avoids storing unnecessary spec/status contents.
2. Controller RBAC: In basically all cases, a controller should be given global get/list/watch its primary objects and update on its primary objects status, so that it is always able to inform users of the state of custom objects across the cluster. However, in many cases, a cluster admin may want to limit permissions of a controller to get/list/watch/create/update/delete secondary objects in specific namespaces. This setup fails in a single global cache scenario because the controller would not have global get/list/watch/create/update/delete permission on secondary resources.

Reply:
Could we maybe clarify the use case in an issue before we dump code? The referenced issue only describes how this works, not what problem it solves. That makes it impossible to provide any useful design feedback.
/hold

Reply:
@alvaroaleman I will update the issue.

Reply:
@alvaroaleman I've updated the issue. 
> We have operators that are moving from namespace scoped to cluster scoped. An operator uses caches for dependent resources, these caches are usually spun up with list/watchers for those resources using controller-runtime pattern at startup.
> 
> If the operator is started with access only to the owned CRDs to start and not given all of the required permissions at startup. The caches will not have the dependent resources and when the admin grants permissions later the caches will not be able to react to the change.
> 
> The operator would need access to a cache that can handle dynamically changing permissions.